### PR TITLE
Add verify_client_cert_signed_by_trust_anchor

### PR DESCRIPTION
**Pull request by iamwwc**
fix https://github.com/rustls/rustls/pull/1134

### CODE REVIEW

**Comment by iamwwc** [comment]
If this PR get merged, please release a new patch version quickly, thank you

File: rustls/src/webpki/verify.rs [hunk 1]
```diff
@@ -65,6 +66,38 @@ pub fn verify_server_cert_signed_by_trust_anchor(
...
+    let webpki_now = webpki::Time::try_from(now).map_err(|_| Error::FailedToGetCurrentTime)?;
+
+    cert.0
+        .verify_for_usage(
+            SUPPORTED_SIG_ALGS,
+            &trust_roots,
+            &chain,
+            webpki_now,
+            webpki::KeyUsage::client_auth(),
+            revocation, // no CRLs
```
Comment (djc) [inline rustls/src/webpki/verify.rs:95]: The `no CRLs` comment doesn't make sense in the client context, where a `Option<RevocationOptions<'_>>` is actually passed in.

(Is `RevocationOptions<'_>` part of the public API?

**Review by djc — COMMENTED** [review]
LGTM. Once this is merged, can you do a backport to the rel-0.21 branch?

**Comment by iamwwc** [comment]
> LGTM. Once this is merged, can you do a backport to the rel-0.21 branch?

I can use cargo `rustls={git=xxx}` temporarily, when the patch version will release?

File: rustls/src/webpki/verify.rs (hunk 1 above)
Comment (ctz) [reply rustls/src/webpki/verify.rs]: `RevocationOptions` is part of the webpki public API

Comment (iamwwc) [reply rustls/src/webpki/verify.rs]: I prefer so, If user don't want to check CRLs, they can pass None

Comment (djc) [reply rustls/src/webpki/verify.rs]: We don't want to have webpki public API in the rustls public API, so we'll need a different solution here.

Comment (iamwwc) [reply rustls/src/webpki/verify.rs]: `WebPkiClientVerifier#new` already rely on `webpki::OwnedCertRevocationList`

**Review by cpu — COMMENTED** [review]
> iamwwc added 3 commits 5 hours ago

Could you rebase these three commits into one? I don't think the comment update and chore commit are valuable on their own.

**Comment by iamwwc** [comment]
I will rebase into one commit when other revision advice is done

**Comment by iamwwc** [comment]
I haveadd the `RustlsRevocationOptions` to avoid leak the `webpki::RevocationOptions` to public api

**Comment by djc** [comment]
Your `RustlsRevocationOptions` has a `From<webpki::RevocationOptions>` impl, which still means the `webpki::RevocationOptions` type becomes part of the public rustls API.

**Comment by iamwwc** [comment]
any advice? I don't know how to solve it gracefully

**Comment by djc** [comment]
There's no easy fix. One option is to push down the transitive set of required types down into the new rustls-pki-types crate, but we can't backport that to 0.21.

**Comment by iamwwc** [comment]
> LGTM. Once this is merged, can you do a backport to the rel-0.21 branch?

I try use main branch code via `rustls={git=xxx}` in my private repo. I found latest type `ClientConfig` changed to `ClientConfig<C>`, but latest `tokio-rustls#TlsConnector#from` does't support `ClientConfig<C>`. How I can do that?

Yes. Once this PR get merged, I will cherry-pick commit and make a new PR to `rel-0.21` branch

**Comment by djc** [comment]
Yeah, there is no tokio-rustls branch to match current rustls main. See #1435 for release planning.

**Comment by ctz** [comment]
I think people wanting to reuse `WebPkiClientVerifier` should wrap it. eg,

```
struct MyVerifier {
  parent: rustls::server::WebPkiClientVerifier,
}

impl rustls::server::ClientCertVerifier for MyVerifier {
  fn verify_client_cert(&self, ...) -> ... {
    let assertion = self.parent.verify_client_cert(...)?;
    // extra validations as desired here
    Ok(assertion)
  }
}
```

The `verify_client_cert_signed_by_trust_anchor` introduced in this PR is the entirety of `WebPkiClientVerifier::verify_client_cert`, the only difference is avoiding parsing the certificate twice, but that is extremely fast and zero-copy.

**Comment by iamwwc** [comment]
> I think people wanting to reuse `WebPkiClientVerifier` should wrap it. eg,
> 
> ```
> struct MyVerifier {
>   parent: rustls::server::WebPkiClientVerifier,
> }
> 
> impl rustls::server::ClientCertVerifier for MyVerifier {
>   fn verify_client_cert(&self, ...) -> ... {
>     let assertion = self.parent.verify_client_cert(...)?;
>     // extra validations as desired here
>     Ok(assertion)
>   }
> }
> ```
> 
> The `verify_client_cert_signed_by_trust_anchor` introduced in this PR is the entirety of `WebPkiClientVerifier::verify_client_cert`, the only difference is avoiding parsing the certificate twice, but that is extremely fast and zero-copy.

Yes, This is a way to solve it. but It's a hack

I currently use another hack to solve it.  But I think we should have a suitable way to solve it

**Comment by djc** [comment]
Why is this a hack?

**Comment by cpu** [comment]
Flipping this to draft status since it seems there's outstanding work to be done.

**Pull request closed without merging**